### PR TITLE
If the rasterizer has a valid surface context, use that to create a snapshotting render target.

### PR DESCRIPTION
**Pull request by chinmaygarde**
Fixes snapshotting of textures uploaded on the IO thread and resident only on the GPU. Does **not** fix snapshotting of scenes with external textures.

Fixes https://github.com/flutter/flutter/issues/16412. The engine refactor removed all the various places the engine used to screenshot layer trees and moved them to a single spot on the rasterizer. This patch fixes that call to handle IO thread uploaded textures.

### CODE REVIEW

File: shell/common/rasterizer.cc [hunk 1]
```diff
@@ -130,27 +131,67 @@ static sk_sp<SkPicture> ScreenshotLayerTreeAsPicture(flow::LayerTree* tree) {
...
   flow::CompositorContext compositor_context;
-  auto canvas = bitmap_surface->getCanvas();
-  auto frame = compositor_context.AcquireFrame(nullptr, canvas, false);
+  auto canvas = snapshot_surface->getCanvas();
+  auto frame = compositor_context.AcquireFrame(surface_context, canvas, false);
   canvas->clear(SK_ColorBLACK);
   frame->Raster(*tree, true);
   canvas->flush();
+
+  // Prepare an image from the surface, this image may potentially be on th GPU.
```
Comment (jason-simmons) [inline shell/common/rasterizer.cc]: typo: "the"

**Comment by cbracken** [comment]
lgtm

**Comment by chinmaygarde** [comment]
Tree is green after the shell refactor. So I am assuming it stuck the landing this time. That unblocks this patch.

**Comment by tvolkert** [comment]
This has been identified as the culprit of a Windows crash on AppVeyor trying to roll this into flutter/flutter.

https://ci.appveyor.com/project/flutter/flutter/build/1.0.11811

```
RUNNING: cd dev\automated_tests; ..\..\bin\flutter.bat test test_smoke_test\pass_test.dart
00:00 +0 -1: loading C:\projects\flutter sdk\dev\automated_tests\test_smoke_test\pass_test.dart [E]
  Failed to load "C:\projects\flutter sdk\dev\automated_tests\test_smoke_test\pass_test.dart":
  Shell subprocess crashed with unexpected exit code -1073741819 before connecting to test harness.
  Test: C:\projects\flutter sdk\dev\automated_tests\test_smoke_test\pass_test.dart
  Shell: C:\projects\flutter sdk\bin\cache\artifacts\engine\windows-x64\flutter_tester.exe
```

**Comment by tvolkert** [comment]
Scratch that - sounds like the culprit may be a different PR

**Comment by andreidiaconu** [comment]
@chinmaygarde While trying to screenshot images, sometimes they do not render. This seems to address a similar issue. Perhaps they are related? Opened issue: https://github.com/flutter/flutter/issues/17687

**Pull request merged**